### PR TITLE
cpplint: Use pypi version

### PR DESCRIPTION
**Pull request by HaoZeke**
## Summary

Python 2 has been depreciated for quite a while now, and it isn't mandated by any other part of the process.

The [fork on pypi](https://pypi.python.org/pypi/cpplint) is probably the one most people will expect anyway, so have simply replaced the existing `cpplint` with that, adding to the ignore rules to ensure no further changes are required at this point.

P.S.: I would actually prefer to follow the flag guidelines, but let me know if that would be too invasive of a change.

## Tests

With the new `cpplint`:

```
Category 'build/header_guard' errors found: 2805
Category 'build/include_order' errors found: 3301
Category 'build/include_what_you_use' errors found: 23
Category 'build/namespaces' errors found: 3
Category 'runtime/string' errors found: 36
Total errors found: 6168
make: *** [make/cpplint:5: cpplint] Error 1
```
With the old one:

```
Total errors found: 0
```

The most common errors are essentially related to ordering of system libraries and the fact that the newer guidelines like the guards to be in a different format, as well as the suggestion of having a C style string for global and static strings:

```
#ifndef STAN_MATH_REV_FUN_VARIANCE_HPP -> #ifndef STAN_MATH_REV_FUN_VARIANCE_HPP_
#endif -> #endif // STAN_MATH_REV_FUN_VARIANCE_HPP_
For a static/global string constant, use a C style string instead: "const char MAJOR_VERSION[]". [4]
stan/math/rev/functor/cvodes_ode_data.hpp(195): error cpplint: [build/include_what_you_use] Add #include <utility> for move [4]
```

Have currently ignored these suggestions (`flag` errors format):
- `build/include_order` 3301
- `build/header_guard` 2805
- `build/namespaces` 3
- `runtime/string` 36
- `build/include_what_you_use` 23

To ensure `make cpplint` does not error out, but am happy to update as per the linter guidelines as well. 

TL;DR- Post PR:

```
Total errors found: 0
```

i.e. no change.

## Side Effects

Will have to make a similar PR to clean up the `make/cpplint` file in `stan-dev/stan` as well.

## Release notes

Updated `cpplint` to conform to `pypi`

## Checklist

- [X] Math issue #(issue number) **N/A**

- [X] Copyright holder: Rohit Goswami <rg0swami[at]yahoo.com>

    The copyright holder is typically you or your assignee, such as a university or company. By submitting this pull request, the copyright holder is agreeing to the license the submitted work under the following licenses:
      - Code: BSD 3-clause (https://opensource.org/licenses/BSD-3-Clause)
      - Documentation: CC-BY 4.0 (https://creativecommons.org/licenses/by/4.0/)

- [x] the basic tests are passing

    - unit tests pass (to run, use: `./runTests.py test/unit`)
    - header checks pass, (`make test-headers`)
    - dependencies checks pass, (`make test-math-dependencies`)
    - docs build, (`make doxygen`)
    - code passes the built in [C++ standards](https://github.com/stan-dev/stan/wiki/Code-Quality) checks (`make cpplint`)

The `doxygen` test has issues (depreciation warnings), but not related to this PR, and not with the version of `doxygen` specified.

- [X] the code is written in idiomatic C++ and changes are documented in the doxygen **N/A**

- [X] the new changes are tested

### CODE REVIEW

**Comment by bob-carpenter** [comment]
Thanks for the patch.  Is this just configuring the Python 3 version to work like our Python 2 version did?  Or does this change code to comply?  I couldn't quite follow from the PR notes.

There's always a copyright holder for code.  That's the author by default, or their employer if they assign copyright to code to their employer (as most employers, including most universities, require).  

We asked explicitly to get a record of copyright holders and to make sure contributors were OK licensing it open source.  I intentionally wanted to keep the contributors as copyright holders so that nobody could try to seize control of the whole project. 

What I think is probably a better approach now is to remind users that they have to have copyright and contribute according to our licensing from the GitHub terms of service, to which we can point in the PR template.  That'd stop me from nagging everyone to fill in the PR template and I think we're still OK on copyright.  This is what @seantalts did proactively for the stanc3 repo, but we haven't adopted that practice in the main math or stan repos.

**Comment by HaoZeke** [comment]
Sorry for the confusion.

This PR just switches to the `python 3` compatible version and adds ignore rules to ensure no changes are required in the code right now. 

I have updated the PR with the relevant copyright.

**Comment by SteveBronder** [comment]
@serban-nicusor-toptal do we need to update the jenkins machines to use python3 as well?

**Comment by serban-nicusor-toptal** [comment]
All the machines should have python3 installed, does this PR require any change to the [Jenkinsfile](https://github.com/stan-dev/math/blob/develop/Jenkinsfile#L6) ?

**Comment by rok-cesnovar** [comment]
No. jenkins calls make cpplint. This changes what happens underneath.

**Comment by serban-nicusor-toptal** [comment]
Good, I'll double-check all machines now for python3 to be sure. I'll comment here when done.

**Comment by rok-cesnovar** [comment]
> I would actually prefer to follow the flag guidelines, but let me know if that would be too invasive of a change.

I agree. I feel that this way of doing it by first replacing cpplint and then applying the changes to follow guidelines will be easier to review.

**Comment by serban-nicusor-toptal** [comment]
Hey, here are the results:

```
AWS Linux - Yes ( python3 resolves to Python 3.5.2 )
AWS Windows - Yes ( python resolves to Python 3.8.2 )
gelman-group-mac - No ( No root access )
gelman-group-win2 - Yes ( python resolves to Python 3.8.2 )
old-imac - Yes ( python3 resolves to Python 3.4.2 )
gelman-group-win-new - Machine down
gelman-group-linux - Machine down
```

Will this create a problem on mac ? I don't have root access to it.

**Comment by rok-cesnovar** [comment]
Restarted the tests so that the upstream Stan branch points to the stan-dev/stan PR.

@serban-nicusor-toptal can we remove the mac machine for cpplint until someone either gives you acess or upgrades to python 3.

**Comment by serban-nicusor-toptal** [comment]
Done

**Comment by rok-cesnovar** [comment]
@serban-nicusor-toptal Is the hold up in this test https://jenkins.mc-stan.org/blue/organizations/jenkins/Stan/detail/PR-2916/3/pipeline/142
because of this change of the mac worker?

**Comment by serban-nicusor-toptal** [comment]
Hey, sadly no. Jenkins will spin up new Linux nodes as the queue is growing, lately, we've had a bit of delay because of almost all machines in AWS being used. ( Note: We are using Spot Instances which are 70-80% cheaper but if someone pays the full price aka on-demand for the instance we lose it )

AWS Logs Example:
 ```
Repeated errors have occurred processing the launch specification "m5dn.4xlarge, ami-0a5df599f96078980, Linux/UNIX, us-east-2b while launching spot instance". It will not be retried for at least 13 minutes. Error message: There is no Spot capacity available that matches your request. (Service: AmazonEC2; Status Code: 500; Error Code: InsufficientInstanceCapacity)
```

How to figure this out without having access to AWS ? 
image
In Jenkins you will see that it requested `target: 2` but `nodes: 0` is still 0 because it doesn't have any machine available.

I will add new types of machines ( a bit more expensive ) to get us more chances for a machine type, tho lately most of the machine types are being depleted.
There's also the option to switch to on-demand instances but the price difference will be quite big.

More information about [Spot Requests](https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/using-spot-instances.html)

**Comment by rok-cesnovar** [comment]
Oh yeah, that makes sense with a lot of people working from home these days. I guess. Thanks for the explanation.

**Comment by rok-cesnovar** [comment]
@HaoZeke congrats on your first merged PR to the Stan Math repository. :tada:

**Comment by HaoZeke** [comment]
Thanks @rok-cesnovar and everyone else. It was fun! Should I start applying more of the default linter rules to the code? Perhaps that should be an issue or a discourse topic, since people who are working with the code right now should be informed or they will suddenly see a lot of broken tests.
